precompiles/erc20/utils_test.go

```diff
@@ -110,17 +119,108 @@ func (s *PrecompileTestSuite) requireSendAuthz(grantee, granter sdk.AccAddress,
 
 	return precompile
 }
+
+// callType constants to differentiate between direct calls and calls through a contract.
+const (
+	directCall = iota + 1
+	contractCall
+)
+
+// getCallArgs is a helper function to return the correct call arguments for a given call type.
+//
+// In case of a direct call to the precompile, the precompile's ABI is used. Otherwise, the
+// ERC20CallerContract's ABI is used and the given contract address.
+func (s *PrecompileTestSuite) getTxAndCallArgs(callType int, contractAddr common.Address) (evmtypes.EvmTxArgs, factory.CallArgs) {
+	txArgs := evmtypes.EvmTxArgs{}
+	callArgs := factory.CallArgs{}
+
+	switch callType {
+	case directCall:
+		precompileAddr := s.precompile.Address()
+		txArgs.To = &precompileAddr
+		callArgs.ContractABI = s.precompile.ABI
+	case contractCall:
+		txArgs.To = &contractAddr
+		callArgs.ContractABI = testdata.ERC20CallerContract.ABI
+	}
+
+	return txArgs, callArgs
+}
+
+// callContractAndCheckLogs is a helper function to call a contract and check the logs using
+// the integration test utilities.
+//
+// It returns the Cosmos Tx response, the decoded Ethereum Tx response and an error. This error value
+// is nil, if the expected logs are found and the VM error is the expected one, should one be expected.
+//
+// TODO: add this to network utils?
+func (s *PrecompileTestSuite) callContractAndCheckLogs(
+	priv cryptotypes.PrivKey,
+	txArgs evmtypes.EvmTxArgs,
+	callArgs factory.CallArgs,
+	logCheckArgs testutil.LogCheckArgs,
+) (abci.ResponseDeliverTx, *evmtypes.MsgEthereumTxResponse, error) {
+	res, err := s.factory.ExecuteContractCall(priv, txArgs, callArgs)
+	logCheckArgs.Res = res
+	if err != nil {
+		// NOTE: here we are still passing the response to the log check function,
+		// because we want to check the logs and expected error in case of a VM error.
+		//
+		// TODO: refactor CheckLogs function
+		return abci.ResponseDeliverTx{}, nil, CheckError(err, logCheckArgs)
+	}
+
+	ethRes, err := evmtypes.DecodeTxResponse(res.Data)
+	if err != nil {
+		return abci.ResponseDeliverTx{}, nil, err
+	}
+
+	return res, ethRes, testutil.CheckLogs(logCheckArgs)
+}
+
+// CheckError is a helper function to check if the error is the expected one.
+//
+// TODO: improve this
+func CheckError(err error, logCheckArgs testutil.LogCheckArgs) error {
+	if (err == nil) && logCheckArgs.ExpPass {
+		return nil
+	}
+
+	if (err == nil) && !logCheckArgs.ExpPass {
+		return errorsmod.Wrap(err, "expected error but got none")
+	}
+
+	if (err != nil) && logCheckArgs.ExpPass {
+		return errorsmod.Wrap(err, "expected no error but got one")
+	}
+
+	if logCheckArgs.ErrContains == "" {
+		// NOTE: if err contains is empty, we return the error as it is
+		return errorsmod.Wrap(err, "ErrContains needs to be filled")
+	}
+
+	if !strings.Contains(err.Error(), logCheckArgs.ErrContains) {
+		return errorsmod.Wrapf(err, "expected different error; wanted %q", logCheckArgs.ErrContains)
+	}
+
+	return nil
```

Comment:
This is not really pretty right now but functional, will probably refactor

Reply:
you can use a `switch` instead to make it cleaner